File: FriTD/src/Entities/Projectile.java
package Entities;

import java.awt.Color;
import java.awt.Graphics2D;

import Enums.ProjectileState;
import Helpers.MathHelper;
import UI.IDisplayableObject;

public class Projectile  implements IDisplayableObject{
	
	private int speed;
	private Effect effect;
	private int x,y;
	private ProjectileState state;
	private Enemy target;
	
	

	public Projectile(int speed,Effect effect, int x, int y, Enemy target) {
		super();
		this.speed = speed;
		this.effect = effect;
		this.x = x;
		this.y = y;
		state = ProjectileState.active;
		this.target = target;
	}

	@Override
	public void display(Graphics2D g2d) {
		Color c = g2d.getColor();
		g2d.setColor(Color.orange);
		g2d.fillOval(x-7, y-7, 15, 15);
		g2d.setColor(c);
		
	}

	public boolean isEmpty() {
		return state == ProjectileState.nonactive;
	}

	public void move() {
		if(ProjectileState.nonactive == state)return;
		int nextX, nextY;
		
		if(target.getX()!=x){
			if(target.getX() < x){
				nextX = x - speed;
			}else{
				nextX = x + speed;
			}
		}else{
			nextX = x;
		}
		if(target.getY()!=y){
			if(target.getY() < y){
				nextY = y - speed;
			}else{
				nextY = y + speed;
			}
		}else{
			nextY = y;
		}
		
		//MCH
		if(MathHelper.getDistanceBetweenPoints(x, y, target.getX(), target.getY())<speed){
			target.applyEffect(effect);
			this.effect = null;
			this.state = ProjectileState.nonactive;
			
		};
		
		x = nextX;
		y = nextY;
		
	}

}

File: FriTD/TD/Entities/Projectile.cs
﻿using TD.Core;
using TD.Enums;
using TD.Helpers;

namespace TD.Entities
{
    public class Projectile : IDisplayableObject
    {
        private static int _nextSeqId; // is zero by default

        private readonly int _speed;
        private readonly Enemy _target;
        private Effect _effect;

        public int X { get; set; }
        public int Y { get; set; }
        public int Id { get; set; }
        public int SeqId { get; set; }
        public double Perc { get; set; }
        public int WX { get; set; }
        public int WY { get; set; }
        public ProjectileState State { get; private set; }
        public TDGame Game { get; set; }

        public Projectile(int speed, Enemy target, Effect effect)
        {
            State = ProjectileState.Active;
            _speed = speed;
            _target = target;
            _effect = effect;
            SeqId = _nextSeqId++;
        }

        public bool IsEmpty()
        {
            return State == ProjectileState.NonActive;
        }

        public void Move()
        {
            if (ProjectileState.NonActive == State) return;
            int nextX, nextY;

            if (_target.X != X)
            {
                if (_target.X < X)
                {
                    nextX = X - _speed;
                }
                else
                {
                    nextX = X + _speed;
                }
            }
            else
            {
                nextX = X;
            }

            if (_target.Y != Y)
            {
                if (_target.Y < Y)
                {
                    nextY = Y - _speed;
                }
                else
                {
                    nextY = Y + _speed;
                }
            }
            else
            {
                nextY = Y;
            }

            //MCH
            if (MathHelper.DistanceBetweenPoints(X, Y, _target.X, _target.Y) < _speed)
            {
                _target.ApplyEffect(_effect);

                if (_effect.IsSplash())
                {
                    GlobalEventHandler.AreaOfDmg(_target.X, _target.Y, _effect.PrimaryDmg, _effect.DmgType, _effect.SplashRadius, Game);
                }

                _effect = null;
                State = ProjectileState.NonActive;
                X = 0;
                Y = 0;
                return;
            }

            X = nextX;
            Y = nextY;
        }
    }
}
